wurst/math/Matrices.wurst

```diff
@@ -0,0 +1,310 @@
+
+/** Flips the matrix over its main diagonale.
+Result of transposing for rotation matrix is its inverse matrix. */
+public function mat2.transpose() returns mat2
+    return mat2(
+        this.rc00, this.rc10,
+        this.rc01, this.rc11)
+
+/** Returns a zero-matrix in case there's no inverse matrix.*/
+public function mat2.inverse() returns mat2
+    var d = this.determinant()
+    if d.abs().isBetween(0.998, 1.002)
+        return this.transpose()
+    else if d.abs() > 0.002
+        var m = mat2(
+            this.rc11, -this.rc01,
+            -this.rc10, this.rc00)
+        return 1/d*m
+    return ZERO22
+
+public function mat2.toString() returns string
+    return "Matrix 2x2\n{0} {1}\n{2} {3}".format(
+        this.rc00.toString(), this.rc01.toString(),
+        this.rc10.toString(), this.rc11.toString())
+
+/** Creates rotation matrix from angle. */
+public function angle.toMatrix() returns mat2
+    return mat2(
+        this.cos(), -this.sin(),
+        this.sin(), this.cos())
+
+/** Creates scaling 2x2 matrix from a vector. */
+public function vec2.toScaling() returns mat2
+    return mat2(
+        this.x, 0,
+        0, this.y)
+    
+/* =========== 3X3 MATRIX =========== */
+/** Matrix 3x3
+00 01 02
+10 11 12
+20 21 22
+*/
+public tuple mat3(
+    real rc00, real rc01, real rc02,
+    real rc10, real rc11, real rc12,
+    real rc20, real rc21, real rc22)
+
+public function mat3.op_plus(mat3 m) returns mat3
+    return mat3(
+        this.rc00 + m.rc00, this.rc01 + m.rc01, this.rc02 + m.rc02,
+        this.rc10 + m.rc10, this.rc11 + m.rc11, this.rc12 + m.rc12,
+        this.rc20 + m.rc20, this.rc21 + m.rc21, this.rc22 + m.rc22)
+
+public function mat3.op_plus(real scalar) returns mat3
+    return mat3(
+        this.rc00 + scalar, this.rc01 + scalar, this.rc02 + scalar,
+        this.rc10 + scalar, this.rc11 + scalar, this.rc12 + scalar,
+        this.rc20 + scalar, this.rc21 + scalar, this.rc22 + scalar)
+
+public function mat3.op_minus(mat3 m) returns mat3
+    return mat3(
+        this.rc00 - m.rc00, this.rc01 - m.rc01, this.rc02 - m.rc02,
+        this.rc10 - m.rc10, this.rc11 - m.rc11, this.rc12 - m.rc12,
+        this.rc20 - m.rc20, this.rc21 - m.rc21, this.rc22 - m.rc22)
+        
+public function mat3.op_minus(real scalar) returns mat3
+    return mat3(
+        this.rc00 - scalar, this.rc01 - scalar, this.rc02 - scalar,
+        this.rc10 - scalar, this.rc11 - scalar, this.rc12 - scalar,
+        this.rc20 - scalar, this.rc21 - scalar, this.rc22 - scalar)
+
+public function mat3.op_mult(real scalar) returns mat3
+    return mat3(
+        this.rc00*scalar, this.rc01*scalar, this.rc02*scalar,
+        this.rc10*scalar, this.rc11*scalar, this.rc12*scalar,
+        this.rc20*scalar, this.rc21*scalar, this.rc22*scalar)
+
+/** Matrix multiplication is not commutative that means matrix*vect is not the same as vec*matrix. */
+public function real.op_mult(mat3 m) returns mat3

```

Comment:
How is this comment relevant to scalar multiplication?

Reply:
Funny, thanks.